### PR TITLE
#2306 Forbid creating/updating a recipe if API version of the extension changes

### LINKED ISSUE #2306: Block automatic v2 to v3 upgrade of extensions that are part of a blueprint with multiple extensions
Discussion here: https://github.com/pixiebrix/pixiebrix-extension/pull/2298#issuecomment-1007003564

### PR DESCRIPTION
**Pull request by BALEHOK**
Resolves #2306

### CODE REVIEW

File: src/devTools/editor/panes/save/SavingExtensionModal.test.tsx [hunk 1]
```diff
@@ -50,41 +67,34 @@ test("renders all buttons when Recipe is editable", () => {
...
 
   const saveAsPersonalExtensionButton = screen.getByText(
     "Save as Personal Extension"
   );
-  expect(saveAsPersonalExtensionButton).toHaveClass("btn-secondary");
+  expect(saveAsPersonalExtensionButton).toHaveClass("btn-primary");
 
   const cancelButton = screen.getByText("Cancel");
   expect(cancelButton).toHaveClass("btn-info");
-});
```
Comment (BALEHOK) [inline src/devTools/editor/panes/save/SavingExtensionModal.test.tsx:80]: This was a duplicate of "renders new recipe button when recipe is not editable".

File: src/devTools/editor/panes/save/SavingExtensionModal.tsx [hunk 2]
```diff
@@ -43,7 +43,13 @@ const SavingExtensionModal: React.FC<OwnProps> = ({
   let showNewRecipeButton = false;
   let showUpdateRecipeButton = false;
 
-  if (isRecipeEditable) {
+  const recipeName = recipe.metadata.name;
+  const installedRecipeVersion = element.recipe.version;
+  const latestRecipeVersion = recipe.metadata.version;
+
+  if (recipe.apiVersion !== element.apiVersion) {
+    message = `This extension is part of blueprint ${recipeName}. The API version ${recipe.apiVersion} of the blueprint is not compatible with the current API version ${element.apiVersion} of the extension.`;
```
Comment (BLoe) [inline src/devTools/editor/panes/save/SavingExtensionModal.tsx:51]: Maybe could also add to the end of this, something like: "You may still save `${element.label}` as a personal extension"

**Review by BLoe — APPROVED** [review]
Add to the message if you think it makes sense, but otherwise LGTM 👍

**Review by BLoe — CHANGES_REQUESTED** [review]
Sorry, I jumped the gun with my approval, just thought of something else.

I was thinking about the message text and thought it was unclear what it was actually referring to, since the button is removed in this case. I wonder if we should instead show the button, but disabled. This might be more clear what's going on to users.

**Comment by twschiller** [comment]
Double-checking my understanding:

1. If the extension is the only extension in the blueprint, we want to save the blueprint with the new version. (It should be safe)
2. If there's multiple extensions in the blueprint, then prevent updating the blueprint if the version doesn't match

Glancing at the code I don't see the logic for 1?

**Comment by BLoe** [comment]
> Glancing at the code I don't see the logic to allow 1?

We definitely forgot to go over this part earlier today, sorry about that. I assume we want to update the `apiVersion` property on the blueprint in this case?

**Comment by twschiller** [comment]
> I assume we want to update the `apiVersion` property on the blueprint in this case?

Yes, we need to update the apiVersion on the blueprint in that case. Between v2 and v3 we didn't change how the `definitions` and `options` sections work, so this should be safe

**Comment by BALEHOK** [comment]
> Sorry, I jumped the gun with my approval, just thought of something else.
> 
> I was thinking about the message text and thought it was unclear what it was actually referring to, since the button is removed in this case. I wonder if we should instead show the button, but disabled. This might be more clear what's going on to users.

Currently we don't show Blueprint buttons if a user can't update the Blueprint. It makes sense to me to follow the same pattern with the API version mismatch. Otherwise, I can change the wizard to always show buttons but disable them, but for all the cases, not just the one related to API version.

File: src/devTools/editor/panes/save/SavingExtensionModal.tsx (hunk 2 above)
Comment (BALEHOK) [reply src/devTools/editor/panes/save/SavingExtensionModal.tsx:51]: Should it say this for other cases as well then?
API version mismatch is no different for the non-editable case for example.

**Comment by BLoe** [comment]
I think it makes sense in this case just because the message text is referring to the user not allowed to do something, but it's not totally clear what they aren't allowed to do, since there is just a save and cancel button. In other cases, if the extension hasn't been packaged into a blueprint, and there's no message about blueprints, then I think it makes sense to hide the button.

File: src/devTools/editor/panes/save/SavingExtensionModal.tsx (hunk 2 above)
Comment (BLoe) [reply src/devTools/editor/panes/save/SavingExtensionModal.tsx:51]: If you want to keep the style similar to the others, you could make it "Do you want to create a personal extension?" on the end of this one. And you could also add similar questions to the end of those other cases if you want to, yeah.

File: src/devTools/editor/panes/save/SavingExtensionModal.tsx [hunk 3]
```diff
@@ -43,9 +43,16 @@ const SavingExtensionModal: React.FC<OwnProps> = ({
...
+  const installedRecipeVersion = element.recipe.version;
+  const latestRecipeVersion = recipe.metadata.version;
+
+  const canUpdateRecipeApiVersion = recipe.extensionPoints.length <= 1;
+  if (recipe.apiVersion !== element.apiVersion && !canUpdateRecipeApiVersion) {
+    message = `This extension is part of blueprint ${recipeName}. The API version ${recipe.apiVersion} of the blueprint is not compatible with the current API version ${element.apiVersion} of the extension.`;
+  } else if (isRecipeEditable) {
     if (installedRecipeVersion === latestRecipeVersion) {
-      message = `This extension is part of blueprint ${recipeName}, do you want to edit the blueprint, or create a personal extension?`;
+      message = `This extension is part of blueprint ${recipeName}.`;
```
Comment (BLoe) [inline src/devTools/editor/panes/save/SavingExtensionModal.tsx:55]: 👍

**Review by BLoe — APPROVED** [review]

**Pull request merged**